convex-core/src/main/java/convex/core/lang/reader/AntlrReader.java

```diff
@@ -471,12 +482,101 @@ public static AList<ACell> readAll(CharStream cs) {
 		ConvexParser parser = new ConvexParser(tokens);
 		parser.removeErrorListeners();
 		ParseTree tree = parser.forms();
-		
+
 		CRListener visitor=new CRListener();
 		ParseTreeWalker.DEFAULT.walk(visitor, tree);
-		
+
 		ArrayList<ACell> top=visitor.popList();
 		return Lists.create(top);
 	}
 
+	public static ACell readOne(java.io.PushbackReader r) {
+		try {
+			CharStream cs = new InteractiveCharStream(r);
+			ConvexLexer lexer=new ConvexLexer(cs);
+			lexer.removeErrorListeners();
+			lexer.setTokenFactory(new CommonTokenFactory(true));
+			TokenStream tokens = new InteractiveTokenStream(lexer);
+			ConvexParser parser = new ConvexParser(tokens);
+			parser.removeErrorListeners();
+
+			ParseTree tree = parser.form();
+
+			r.unread(cs.LA(1));
```

Comment:
Even after disabling the `sync`ing in the streams with our custom `consume` implementations below, there is a single additional character that is read from the stream when a form is parsed. I'm currently unsure of the exact mechanism that causes this but I believe it involves `consume` calls in antlr's [LexerATNSimulator class](https://github.com/antlr/antlr4/blob/master/runtime/Java/src/org/antlr/v4/runtime/atn/LexerATNSimulator.java).  It's possible that there's a better way to handle this there. But just unreading from the stream here seems to work well.